cloudpickle/cloudpickle_fast.py

```diff
@@ -0,0 +1,777 @@
+    # reconstructor, because using save_reduce semantics prevents us to perform
+    # any other operation than state updating after obj is created.
+
+    # I may encounter reference cycles, although there seems to be checks
+    # preventing this to happen.
+    if "_abc_impl" in clsdict:
+        (registry, _, _, _) = abc._get_dump(obj)
+        clsdict["_abc_impl"] = [
+            subclass_weakref() for subclass_weakref in registry
+        ]
+
+    # On PyPy, __doc__ is a readonly attribute, so we need to include it in
+    # the initial skeleton class.  This is safe because we know that the
+    # doc can't participate in a cycle with the original class.
+    type_kwargs = {"__doc__": clsdict.pop("__doc__", None)}
+
+    if hasattr(obj, "__slots__"):
+        type_kwargs["__slots__"] = obj.__slots__
+        # Pickle string length optimization: member descriptors of obj are
+        # created automatically from obj's __slots__ attribute, no need to
+        # save them in obj's state
+        if isinstance(obj.__slots__, str):
+            clsdict.pop(obj.__slots__)
+        else:
+            for k in obj.__slots__:
+                clsdict.pop(k, None)
+
+    # If type overrides __dict__ as a property, include it in the type kwargs.
+    # In Python 2, we can't set this attribute after construction.
+    # XXX: removed special handling of __dict__ for python2
+    __dict__ = clsdict.pop("__dict__", None)
+    if isinstance(__dict__, property):
+        type_kwargs["__dict__"] = __dict__
+        __dict__ = None
+
+    return (
+        type(obj),
+        (obj.__name__, obj.__bases__, type_kwargs),
+        (clsdict, {}),
+        None,
+        None,
+        class_setstate,
+    )
+
+
+def class_reduce(obj):
+    """Select the reducer depending on the dynamic nature of the class obj"""
+    # XXX: there used to be special handling for NoneType, EllipsisType and
+    # NotImplementedType. As for now this module handles only python3.8+, this
+    # code has been removed.
+    if obj.__module__ == "__main__":
+        return dynamic_class_reduce(obj)
+
+    try:
+        # All classes are caught in this function: pickleable classes are
+        # filtered out by creating a Pickler with no custom class reducer
+        # (thus, falling back to save_global). If it fails to save obj, then
+        # obj is either a non-pickleable builtin or dynamic.
+        pickle.dumps(obj)
+    except Exception:
+        # XXX: previously, we also looked for the __builtin__ module, but this
+        # is python 2 specific.
+        if obj.__module__ == "builtins":
+            if obj in _BUILTIN_TYPE_NAMES:
+                return builtin_type_reduce(obj)
+
+        typ = type(obj)
+        if typ is not obj and isinstance(obj, type):  # noqa: E721
+            return dynamic_class_reduce(obj)
+
+    else:
+        # if pickle.dumps worked out fine, then simply pickle by attribute
+        return NotImplementedError
```

Comment:
Maybe make it more explicit that raising `NotImplementedError` will fallback to the parent pickle class.


```python
# if pickle.dumps worked out fine, then simply fallback to the traditional pickle by attribute
# implemented in the builtin `Pickler.save_global`.
```